packages/ui/src/components/ToggleSection/ToggleSection.tsx

```diff
@@ -0,0 +1,294 @@
+ * the section is 'open'
+ */
+const OpenContent = React.forwardRef<
+  HTMLDivElement,
+  React.HTMLProps<HTMLDivElement>
+>(({ children, ...props }, forwardedRef) => {
+  const context = useToggleSectionContext();
+  const id = composeId(NAME.OPEN_CONTENT, context?.contentId);
+
+  return (
+    <div
+      id={id}
+      ref={forwardedRef}
+      aria-selected={context?.open || false}
+      {...props}
+    >
+      {context?.open ? children : null}
+    </div>
+  );
+});
+
+/**
+ * A toggle that opens and closes
+ * the section
+ *
+ * SEE: https://www.radix-ui.com/docs/primitives/components/toggle
+ */
+const Trigger = React.forwardRef<
+  React.ElementRef<typeof TogglePrimitive.Root>,
+  Omit<React.ComponentPropsWithoutRef<typeof TogglePrimitive.Root>, "asChild">
+>(({ children, onPressedChange, ...toggleProps }, forwardedRef) => {
+  const context = useToggleSectionContext();
+  const controls = composeControls(context?.contentId);
+
+  const handleOnPressedChange = (newPressed: boolean) => {
+    context?.onOpenToggle?.();
+
+    onPressedChange?.(newPressed);
+  };
+
+  return (
+    <TogglePrimitive.Root
+      ref={forwardedRef}
+      asChild
+      aria-controls={controls}
+      pressed={context?.open || false}
+      onPressedChange={handleOnPressedChange}
+      {...toggleProps}
+    >
+      {children}
+    </TogglePrimitive.Root>
+  );
+});
+
+interface ToggleProps extends React.HTMLAttributes<HTMLElement> {
+  children?: React.ReactNode;
+  /** Determine if this toggle opens or closes the section */
+  open: boolean;
+}
+
+const Toggle = React.forwardRef<HTMLElement, ToggleProps>(
+  ({ onClick, open, ...props }, forwardedRef) => {
+    const context = useToggleSectionContext();
+    const controls = composeControls(context?.contentId);
+
+    const handleClick: React.MouseEventHandler<HTMLButtonElement> = (event) => {
+      if (onClick) {
+        onClick(event);
+      }
+
+      context?.onOpenChange?.(open);
+    };
+
+    return (
+      <Slot
+        onClick={handleClick}
+        ref={forwardedRef}
+        aria-controls={controls}
+        {...props}
+      />
+    );
+  },
+);
+
+/**
+ * Generic wrapper that attaches a click
+ * handler to a button that 'opens' the section
+ */
+const Open = React.forwardRef<HTMLElement, Omit<ToggleProps, "open">>(
+  (props, forwardedRef) => {
+    return <Toggle open ref={forwardedRef} {...props} />;
+  },
+);
+
+/**
+ * Generic wrapper that attaches a click
+ * handler to a button that 'closes' the section
+ */
+const Close = React.forwardRef<HTMLElement, Omit<ToggleProps, "open">>(
+  (props, forwardedRef) => {
+    return <Toggle open={false} ref={forwardedRef} {...props} />;
+  },
+);
+
+interface HeaderProps extends HeadingProps {
+  /** The toggle for the component (appears on right side of header) */
+  toggle: React.ReactElement<typeof Toggle>;
+}
+
+/**
+ * A styled header for the section
+ */
+const Header = React.forwardRef<HTMLHeadingElement, HeaderProps>(
```

Comment:
I wonder if including things like a header and icon are a bit too specific for a reusable component?  Is open/closed even too specific?  I wonder about maybe having this really generic bare compoent with primary/secondary states and then moving the styling to more focused InlineEditableForm component on top of it?  Feel free to disagree.  :laughing:  The InlineEditableForm might have support for the error alert, too?

Reply:
Yeah, I was having a problem with making it generic enough while providing some built in stuff that would always be there. This is a totally optional part. You can compose this component excluding the header. Though, if we think this is too much it can be removed and implemented on a as needed basis as well.

Reply:
Well, let's leave it as is - this might be premature over-engineering anyway.  I wouldn't be surprised if we need to split this in half later but let's not worry about it for now.